# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/script/ScriptLoader.cpp
# Commit: 75ae330e1b90
# Full Hash: 75ae330e1b90273c1eafb42b21098b3753e4954a
# Author: Yoshi Cheng-Hao Huang <allstars.chh@gmail.com>
# Date: 2025-09-24 18:03 +0000
# Regressor Bug: 1968890
# File Overlap Count: 3
# Overlapping Files: dom/worklet/WorkletFetchHandler.cpp, dom/workers/ScriptLoader.cpp, dom/script/ScriptLoader.cpp
# Description:
#   Bug 1968890 - Part 6: ScriptLoaders update. r=jonco,dom-worker-reviewers,edenchuang
# ==============================================================================

diff -r c48ad530cfe2 -r 75ae330e1b90 dom/script/ScriptLoader.cpp
--- a/dom/script/ScriptLoader.cpp	Wed Sep 24 18:03:30 2025 +0000
+++ b/dom/script/ScriptLoader.cpp	Wed Sep 24 18:03:30 2025 +0000
@@ -4220,26 +4220,35 @@
     mDocument->AddBlockedNodeByClassifier(cont);
   }
 
+  bool wasHandled = false;
+
+  // A ModuleLoadRequest will be stored either in mDeferRequests or
+  // mLoadingAsyncRequests, but the onerror handler should be triggered later in
+  // ProcessRequests, so we handle ModuleLoadRequest before mDeferRequestrs and
+  // mLoadingAsyncRequests.
   if (aRequest->IsModuleRequest()) {
     MOZ_ASSERT(!aRequest->GetScriptLoadContext()->mIsInline);
-    aRequest->AsModuleRequest()->OnFetchComplete(aResult);
-  }
-
-  if (aRequest->GetScriptLoadContext()->mInDeferList) {
-    MOZ_ASSERT_IF(aRequest->IsModuleRequest(),
-                  aRequest->AsModuleRequest()->IsTopLevel());
+    wasHandled = true;
+
+    ModuleLoadRequest* modReq = aRequest->AsModuleRequest();
+    modReq->OnFetchComplete(aResult);
+
+    MOZ_ASSERT(modReq->IsErrored());
+  } else if (aRequest->GetScriptLoadContext()->mInDeferList) {
+    wasHandled = true;
     if (aRequest->isInList()) {
       RefPtr<ScriptLoadRequest> req = mDeferRequests.Steal(aRequest);
       FireScriptAvailable(aResult, req);
     }
   } else if (aRequest->GetScriptLoadContext()->mInAsyncList) {
-    MOZ_ASSERT_IF(aRequest->IsModuleRequest(),
-                  aRequest->AsModuleRequest()->IsTopLevel());
+    wasHandled = true;
     if (aRequest->isInList()) {
       RefPtr<ScriptLoadRequest> req = mLoadingAsyncRequests.Steal(aRequest);
       FireScriptAvailable(aResult, req);
     }
-  } else if (aRequest->GetScriptLoadContext()->mIsNonAsyncScriptInserted) {
+  }
+
+  if (aRequest->GetScriptLoadContext()->mIsNonAsyncScriptInserted) {
     if (aRequest->isInList()) {
       RefPtr<ScriptLoadRequest> req =
           mNonAsyncExternalScriptInsertedRequests.Steal(aRequest);
@@ -4251,25 +4260,12 @@
       FireScriptAvailable(aResult, req);
     }
   } else if (aRequest->GetScriptLoadContext()->IsPreload()) {
-    if (aRequest->IsModuleRequest()) {
-      aRequest->Cancel();
-    }
     if (aRequest->IsTopLevel()) {
       // Request may already have been removed by
       // CancelAndClearScriptLoadRequests.
       mPreloads.RemoveElement(aRequest, PreloadRequestComparator());
     }
     MOZ_ASSERT(!aRequest->isInList());
-  } else if (aRequest->IsModuleRequest()) {
-    ModuleLoadRequest* modReq = aRequest->AsModuleRequest();
-    if (modReq->IsDynamicImport()) {
-      if (aRequest->isInList()) {
-        modReq->CancelDynamicImport(aResult);
-      }
-    } else {
-      MOZ_ASSERT(!modReq->isInList());
-      modReq->Cancel();
-    }
   } else if (mParserBlockingRequest == aRequest) {
     MOZ_ASSERT(!aRequest->isInList());
     mParserBlockingRequest = nullptr;
@@ -4286,7 +4282,7 @@
     FireScriptAvailable(aResult, aRequest);
     ContinueParserAsync(aRequest);
     mCurrentParserInsertedScript = oldParserInsertedScript;
-  } else {
+  } else if (!wasHandled) {
     // This happens for blocking requests cancelled by ParsingComplete().
     // Ignore cancellation status for link-preload requests, as cancellation can
     // be omitted for them when SRI is stronger on consumer tags.